### PR TITLE
superenv: Fix path removal bug

### PR DESCRIPTION
**Pull request by iMichka**

### CODE REVIEW

File: Library/Homebrew/shims/super/cc [hunk 1]
```diff
@@ -230,16 +230,16 @@ class Cmd
...
     elsif path.start_with?(prefix, tmpdir)
       true
     elsif path.start_with?("/opt/local", "/opt/boxen/homebrew", "/opt/X11", "/sw", "/usr/X11")
       # ignore MacPorts, Boxen's Homebrew, X11, fink
       false
+    # first two paths: reject references to Cellar or opt paths
+    # for unspecified dependencies
+    elsif path.start_with?(cellar) || path.start_with?(opt)
+      dep = path[keg_regex, 2]
+      dep && @deps.include?(dep))
```
Comment (maxim-belkin) [inline Library/Homebrew/shims/super/cc]: `))` -> `)`

File: Library/Homebrew/shims/super/cc [hunk 2]
```diff
@@ -230,16 +230,16 @@ class Cmd
...
-      dep = path[keg_regex, 2]
-      dep && @deps.include?(dep)
     elsif path.start_with?(prefix, tmpdir)
       true
     elsif path.start_with?("/opt/local", "/opt/boxen/homebrew", "/opt/X11", "/sw", "/usr/X11")
       # ignore MacPorts, Boxen's Homebrew, X11, fink
       false
+    # first two paths: reject references to Cellar or opt paths
+    # for unspecified dependencies
+    elsif path.start_with?(cellar) || path.start_with?(opt)
```
Comment (maxim-belkin) [inline Library/Homebrew/shims/super/cc]: why not `path.start_with?(cellar, opt)` ?

Comment (iMichka) [reply Library/Homebrew/shims/super/cc]: I can do that in a second commit maybe? This change is a stylistic change which would have nothing to do with the pull request.

**Review by sjackman — APPROVED** [review]

**Comment by sjackman** [comment]
Thanks, Michka!

**Review by maxim-belkin — APPROVED** [review]
👍

**Comment by sjackman** [comment]
Docker Hub is currently not building images for us. I'll fix it next week. Just a warning, in case merging this doesn't then immediately get used for building bottles in Linuxbrew/core.
https://hub.docker.com/r/linuxbrew/linuxbrew/builds/

**Comment by iMichka** [comment]
OK. I will merge this and build some bottles right away to see if it does not break other things. I even hope it will fix some other weird problems I have been seing lately.

**Comment by xu-cheng** [comment]
Repost from 
https://github.com/Linuxbrew/brew/commit/163340643353646b8d7d365ca7eff842057c4109#commitcomment-30046109

This change may not be desired.

If I understand correctly, `path.start_with?(cellar) || path.start_with?(opt)` becomes dead code after this change. Noted that the previous `path.start_with?(prefix, tmpdir)` will always return true if the path is a cellar or opt path.

**Comment by iMichka** [comment]
@xu-cheng I understand what you mean, and you a probably right. The fact the `path.start_with?(prefix)` will always return true for paths inside the prefix will cause us to catch whatever is installed in the prefix, without having declared the dependency explicitly in the formula.

I re-rested `brew postinstall perl` and here is how the variables are defined for the path that is being wrongly removed:
```
path : /home/linuxbrew/.linuxbrew/Cellar/perl/5.28.0/lib/perl5/5.28.0/x86_64-linux-thread-multi/CORE
formula_prefix : empty
prefix : /home/linuxbrew/.linuxbrew
```

i think that if `formula_prefix` was not empty it would work. The formula parameter is not passed down to `with_build_environment`, which is called in the `perl.rb` formula. I can fix this by:
1) Pass formula here: https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/extend/ENV.rb#L19-L24
2) Pass `self` to `with_build_environment` in the `perl.rb` formula.

**Comment by sjackman** [comment]
@iMichka How about reverting this PR, and adding the keg of this formula to the allowed white list of locations under Cellar and opt that are explicitly allowed?

**Comment by iMichka** [comment]
I opened a pull request to revert these changes. I need to better understand this code.

Why is `with_build_environment` needed in postinstall? It looks like only the `perl` formula uses such thing, and it probably got broken when they added the formula parameters to all these functions in ENV.

**Comment by iMichka** [comment]
I think the logic for the postinstall step would be that perl should depend on perl :D

**Pull request merged**